### PR TITLE
fix(ios): border + corner rendering

### LINKED ISSUE #9208: [iOS] Border on iOS doesn't look like on UWP or Android for rounded corners
### Current behavior

When the BorderThickness is set to 2 or higher, and CornerRadius is set to half of the Height then the Border on iOS looks differently then on UWP and Android:

UWP:
image

AND:
image

iOS:
image

### Expected behavior

Border looks same on all platforms

### How to reproduce it (as minimally and precisely as possible)

<StackPanel HorizontalAlignment="Center"
					Spacing="8"
					VerticalAlignment="Center">

			<Border BorderThickness="2"
					BorderBrush="#FF6096E6"
					HorizontalAlignment="Left"
					Height="28"
					Padding="8,0"
					CornerRadius="14">
				<TextBlock Text="Test Badge"
						   VerticalAlignment="Center"
						   HorizontalAlignment="Center"
						   Foreground="#FF6096E6"
						   FontWeight="Bold"
						   FontSize="15"
						   LineHeight="19" />
			</Border>

			<Border BorderThickness="4"
					BorderBrush="#FF6096E6"
					HorizontalAlignment="Left"
					Height="28"
					Padding="8,0"
					CornerRadius="14">
				<TextBlock Text="Test Badge"
						   VerticalAlignment="Center"
						   HorizontalAlignment="Center"
						   Foreground="#FF6096E6"
						   FontWeight="Bold"
						   FontSize="15"
						   LineHeight="19" />
			</Border>
		</StackPanel>

### Workaround

<Grid HorizontalAlignment="Left"
				  Height="28"
				  Background="#FF6096E6"
				  Padding="2"
				  CornerRadius="14">
				<Grid HorizontalAlignment="Center"
					  VerticalAlignment="Center"
					  Height="24"
					  Background="{ThemeResource ApplicationPageBackgroundThemeBrush}"
					  CornerRadius="12">
					<TextBlock Text="Test Badge"
							   VerticalAlignment="Center"
							   HorizontalAlignment="Center"
							   Foreground="#FF6096E6"
							   FontWeight="Bold"
							   FontSize="15"
							   Padding="6,0"
							   LineHeight="19" />
				</Grid>
			</Grid>

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia, Uno.UI.RemoteControl / Uno.WinUI.RemoteControl

### NuGet package version(s)

4.3.8

### Affected platforms

iOS

### IDE

Visual Studio 2019

### IDE version

Version 16.11.8

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): closes #9208

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

iOS border can only be drawn with perfect circles, not ellipses.

## What is the new behavior?

iOS border should support most of the scenarios.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information
re: #5440, #6891

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Windows.Foundation;
+using Windows.UI.Xaml;
+using Uno.UI;
+
+namespace Windows.UI.Xaml.Shapes;
+
+#if DEBUG
+public partial class BorderLayerRenderer // to remove
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.cs:11]: to remove

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs [hunk 2]
```diff
@@ -29,7 +32,7 @@
 
 namespace Windows.UI.Xaml.Shapes
 {
-	internal class BorderLayerRenderer
+	partial class BorderLayerRenderer
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs:35]: needs `partial` on BorderLayerRenderer.other-platforms.cs as well

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs [hunk 3]
```diff
@@ -487,9 +530,9 @@ private static void CreateGradientBrushLayers(CGRect fullArea, CGRect insideArea
 			sublayers.Add(gradientContainerLayer);
 		}
 
-		private class LayoutState : IEquatable<LayoutState>
+		public class LayoutState : IEquatable<LayoutState>
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs:533]: private

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs [hunk 4]
```diff
@@ -373,45 +353,108 @@ private static IDisposable InnerCreateLayer(UIElement owner, CALayer parent, Lay
...
+				if (cornerBbox.Width > 0 && cornerBbox.Height > 0)
+				{
+					// A1───2B
+					// 0     3
+					// 7     4
+					// D6───5C
+					var pCorner = GetCorner(cornerBbox, corner); // 1234
+					var pNextMid = GetRelativePoint(cornerBbox, numpadDirection: new[] { 8, 6, 2, 4 }[(int)corner]); // mid point of A1, B3, C5, D7
+
+					// given that ASD can only draw arc of a circle (with equal width, height, and diameter),
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs:401]: #define ASD AddLineToPoint

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs [hunk 5]
```diff
@@ -517,17 +560,17 @@ private class LayoutState : IEquatable<LayoutState>
 				BackgroundImage = backgroundImage;
 			}
 
-			public override int GetHashCode()
-				=> (Background?.GetHashCode() ?? 0 + BorderBrush?.GetHashCode() ?? 0) + (int)BackgroundSizing;
+			public override int GetHashCode() => new object[] { Bbox, Background, BackgroundSizing, BorderBrush, CornerRadius, BorderThickness, BackgroundImage }
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs:563]: This looks expensive :) using a xors is likely to be enough.

Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs]: i wonder if the ValueTuple version be good enough?: (Bbox, Background,... ).GetHashCode()
or the anonymous class version?

**Comment by Xiaoy312** [comment]
edit: undone renaming to make diff cleaner for now

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs (hunk 5 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs]: It could be that making a record of it would ease the writing.

**Comment by jeromelaban** [comment]
This change seems odd: https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=51976&view=ms.vss-test-web.build-test-results-tab&runId=1910422&resultId=100409&paneView=debug Looks like it's adding one line at the bottom. Is it expected?

**Comment by Xiaoy312** [comment]
Definitely not... It went from 8,8,8,8 to 8,8,8,9.

**Review by MartinZikmund — APPROVED** [review]

**Pull request merged**